### PR TITLE
Implement PaC pruning on Component deletion

### PR DESCRIPTION
**Pull request by mmorhun**
Signed-off-by: Mykola Morhun <mmorhun@redhat.com>

This PR:
 - Adds functionality that submits a PR into Components repository when it gets deleted. The PR deletes Pipelines as Code definitions related to the Component. Also removes repository webhook if used. Works for both supported git providers: GitHub and GitLab.
 - Introduces a new finalizer `pac.component.appstudio.openshift.io/finalizer` in order to do Pipelines as Code clean up on Component deletion.
 - Splits Pipelines as Code and initial build logic into separate files

### CODE REVIEW

File: controllers/component_build_controller.go [hunk 1]
```diff
@@ -226,28 +209,43 @@ func (r *ComponentBuildReconciler) Reconcile(ctx context.Context, req ctrl.Reque
...
 		}
+
+		// Update PaC annotation
 		if len(component.Annotations) == 0 {
 			component.Annotations = make(map[string]string)
 		}
 		component.Annotations[PaCProvisionAnnotationName] = pacAnnotationValue
+
+		// Add finalizer to clean up Pipelines as Code configuration on component deletion
+		if component.ObjectMeta.DeletionTimestamp.IsZero() {
```
Comment (tkdchen) [inline controllers/component_build_controller.go:242]: Should be `! component.ObjectMeta.DeletionTimestamp.IsZero` here?

Comment (tkdchen) [inline controllers/component_build_controller.go:242]: Another thought, is it worth to make a separate function to examine whether a resource is being deleted? Something like:
```go
func isResourceBeingDeleted() bool {
...
}
if isResourceBeingDeleted(component) {

}
```

Comment (mmorhun) [reply controllers/component_build_controller.go:242]: It's right, we add finalizer only if deletion has not been requested

Comment (mmorhun) [reply controllers/component_build_controller.go:242]: We can do that, but the construction above is a commonly used pattern

File: controllers/component_build_controller.go [hunk 2]
```diff
@@ -203,6 +169,25 @@ func (r *ComponentBuildReconciler) Reconcile(ctx context.Context, req ctrl.Reque
...
+
+		// Delete Pipelines as Code provision finalizer
+		controllerutil.RemoveFinalizer(&component, PaCProvisionFinalizer)
+		if err := r.Client.Update(ctx, &component); err != nil {
+			return ctrl.Result{}, err
+		}
+		log.Info("PaC finalizer removed")
+
+		// Try to clean up Pipelines as Code configuration
+		r.UndoPaCProvisionForComponent(ctx, &component)
```
Comment (Michkov) [inline controllers/component_build_controller.go:186]: why to try to UndoPaCProvisionForComponent on nonPaC component?

Comment (mmorhun) [reply controllers/component_build_controller.go:186]: good catch. Fixed.

**Comment by psturc** [comment]
I can confirm that the CI failures in this PR are related to the issues with CI we were hitting for past couple of days. I've tried to run the build tests locally against cluster-bot cluster with stonesoup that uses this branch of build-service and didn't hit any issues.

**Review by Michkov — APPROVED** [review]

**Pull request merged**